datafusion/src/physical_plan/file_format/file_stream.rs

```diff
@@ -0,0 +1,274 @@
+    remain: Option<usize>,
+    /// A closure that takes a reader and an optional remaining number of lines
+    /// (before reaching the limit) and returns a batch iterator. If the file reader
+    /// is not capable of limiting the number of records in the last batch, the file
+    /// stream will take care of truncating it.
+    file_reader: F,
+}
+
+impl<F> FileStream<F>
+where
+    F: FnMut(Box<dyn Read + Send + Sync>, &Option<usize>) -> BatchIter
+        + Send
+        + Unpin
+        + 'static,
+{
+    pub fn new(
+        object_store: Arc<dyn ObjectStore>,
+        files: Vec<PartitionedFile>,
+        file_reader: F,
+        schema: SchemaRef,
+        limit: Option<usize>,
+    ) -> Self {
+        let read_iter = files.into_iter().map(move |f| -> DataFusionResult<_> {
+            object_store
+                .file_reader(f.file_meta.sized_file)?
+                .sync_reader()
+        });
+
+        Self {
+            file_iter: Box::new(read_iter),
+            batch_iter: Box::new(iter::empty()),
+            remain: limit,
+            schema,
+            file_reader,
+        }
+    }
+
+    /// Acts as a flat_map of record batches over files.
+    fn next_batch(&mut self) -> Option<ArrowResult<RecordBatch>> {
+        match self.batch_iter.next() {
+            Some(batch) => Some(batch),
+            None => match self.file_iter.next() {
+                Some(Ok(f)) => {
+                    self.batch_iter = (self.file_reader)(f, &self.remain);
+                    self.next_batch()
+                }
+                Some(Err(e)) => Some(Err(ArrowError::ExternalError(Box::new(e)))),
+                None => None,
+            },
+        }
+    }
+}
+
+impl<F> Stream for FileStream<F>
+where
+    F: FnMut(Box<dyn Read + Send + Sync>, &Option<usize>) -> BatchIter
+        + Send
+        + Unpin
+        + 'static,
+{
+    type Item = ArrowResult<RecordBatch>;
+
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        _cx: &mut Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        // check if finished or no limit
+        match self.remain {
+            Some(r) if r == 0 => return Poll::Ready(None),
+            None => return Poll::Ready(self.get_mut().next_batch()),
+            Some(r) => r,
+        };
+
+        Poll::Ready(match self.as_mut().next_batch() {
+            Some(Ok(item)) => {
+                if let Some(remain) = self.remain.as_mut() {
+                    if *remain >= item.num_rows() {
+                        *remain -= item.num_rows();
+                        Some(Ok(item))
+                    } else {
+                        let len = *remain;
+                        *remain = 0;
+                        Some(Ok(RecordBatch::try_new(
+                            item.schema(),
+                            item.columns()
+                                .iter()
+                                .map(|column| column.slice(0, len))
+                                .collect(),
```

Comment:
Even in plans when `limit` is pushed down to the `TableProvider` scan, there is still at least one `LimitExec` above it. Thus I think it is likely fine to avoid slicing up the record batches here (though also as you point out, it also likely won't hurt)